src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractSyntaxFactsService.cs

```diff
@@ -431,5 +431,20 @@ public string GetBannerText(SyntaxNode documentationCommentTriviaSyntax, int ban
             => DocumentationCommentService.GetBannerText(documentationCommentTriviaSyntax, bannerLength, cancellationToken);
 
         protected abstract IDocumentationCommentService DocumentationCommentService { get; }
+
+        protected bool IsNodeOrAnyAncestorLeftSideOfDot(SyntaxNode node, ISyntaxFactsService syntaxFacts)
+        {
+            if (syntaxFacts.IsLeftSideOfDot(node))
+            {
+                return true;
+            }
+
+            if (syntaxFacts.IsRightSideOfQualifiedName(node) || syntaxFacts.IsNameOfMemberAccessExpression(node))
+            {
+                return syntaxFacts.IsLeftSideOfDot(node.Parent);
+            }
+
+            return false;
+        }
```

Comment:
this should not be done thi way.  if you really want this to be a generalized algorithm that is available in multiple places (which i don't think is true, and i think it shoudl just be in the single placethat calls this), then make this an ISYntaxFactsExtension.  

The abstract class is for stateful stuff.